ref_ts/views.py

```diff
@@ -224,32 +225,78 @@ def download_refts_resource_files(request, shortkey, *args, **kwargs):
         response_irods = download_bag_from_irods(request, path, use_async=False)
 
         tempdir = tempfile.mkdtemp()
-        bag_save_to_path = tempdir + "/" + str(shortkey) + ".zip"
+        response = assemble_refts_bag(shortkey, response_irods.streaming_content,
+                                      temp_dir=tempdir)
+
+        return response
+    except Exception as e:
+        logger.exception("download_resource_files: %s" % (e.message))
+        response = HttpResponse(status=503)
+        response.content = "<h3>Failed to download this resource!</h3>"
+        return response
+    finally:
+        if tempdir is not None:
+           shutil.rmtree(tempdir)
 
-        with open(bag_save_to_path, 'wb+') as f:
-            for chunk in response_irods.streaming_content:
-                f.write(chunk)
 
-        res_files_fp_arr = ts_utils.generate_resource_files(shortkey, tempdir)
+@api_view(['GET'])
+def rest_download_refts_resource_files(request, shortkey, *args, **kwargs):
+    tempdir = None
+    _, authorized, _ = authorize(request, shortkey,
+                                 needed_permission=ACTION_TO_AUTHORIZE.VIEW_RESOURCE,
+                                 raises_exception=True)
+    try:
 
-        bag_zip_obj = zipfile.ZipFile(bag_save_to_path, "a", zipfile.ZIP_DEFLATED)
-        bag_content_base_folder = str(shortkey) + "/data/contents/" # _RESOURCE_ID_/data/contents/
-        for fn_fp in res_files_fp_arr:
-            fh = open(fn_fp['fullpath'], 'r')
-            bag_zip_obj.writestr(bag_content_base_folder + fn_fp['fname'], fh.read())
-            fh.close()
-        bag_zip_obj.close()
+        path = "bags/" + str(shortkey) + ".zip"
+        response_irods = download_bag_from_irods(request, path, rest_call=True, use_async=False)
 
-        response = FileResponse(open(bag_save_to_path, 'rb'), content_type='application/zip')
-        response['Content-Disposition'] = 'attachment; filename="' + str(shortkey) + '.zip"'
-        response['Content-Length'] = os.path.getsize(bag_save_to_path)
+        if not response_irods.streaming:
+            raise Exception("Failed to stream RefTS bag")
+        else:
+            tempdir = tempfile.mkdtemp()
+            response = assemble_refts_bag(shortkey, response_irods.streaming_content,
+                                          temp_dir=tempdir)
+            return response
 
-        return response
     except Exception as e:
-        logger.exception("download_resource_files: %s" % (e.message))
+        logger.exception("rest_download_resource_files: %s" % (e.message))
         response = HttpResponse(status=503)
```

Comment:
@zhiyuli good point. Now looking at the rest client code it seems we should not return 200 when something goes wrong. I was only looking at the other download view function. So keep the 503. Can you then change the status code for the other download view function not to return 200? That's in Django_irods though.